### PR TITLE
Feat: #437 remove from categories(folder or tag)

### PR DESCRIPTION
**Pull request by Rccubc**
#437

### CODE REVIEW

File: app/models/categorizer.ts [hunk 1]
```diff
@@ -110,3 +110,11 @@ export enum CategorizerType {
   PaperTag = "PaperTag",
   PaperFolder = "PaperFolder",
 }
+
+export type CategorizeListItem = {
```
Comment (GeoffreyChen777) [inline app/models/categorizer.ts:114]: I would suggest move these to `contextmenu-service.ts` as they are only used there.

File: app/renderer/ui/main-view/data-view/paper-data-view.vue [hunk 2]
```diff
@@ -147,11 +154,39 @@ disposable(
   )
 );
 
+const getCategorizeList = (indexList: number[]) => {
```
Comment (GeoffreyChen777) [inline app/renderer/ui/main-view/data-view/paper-data-view.vue:157]: rename `indexList` with `selectedIndex`.

File: app/renderer/ui/main-view/data-view/paper-data-view.vue [hunk 3]
```diff
@@ -147,11 +154,39 @@ disposable(
...
+    paper.tags.forEach((tag) => {
+      categorizeMap.set(`${tag._id}`, {
+        type: CategorizerType.PaperTag,
+        name: tag.name,
+        id: tag._id,
+      } as CategorizeListItem);
+    });
+  });
+
+  return Array.from(categorizeMap.values()).sort(
```
Comment (GeoffreyChen777) [inline app/renderer/ui/main-view/data-view/paper-data-view.vue:177]: I would suggest to test the performance if we have numerous categorizers. Say if we have 50 tags.

File: app/renderer/ui/main-view/main-view.vue [hunk 4]
```diff
@@ -9,6 +9,8 @@ import { PaperEntity } from "@/models/paper-entity";
 import { IFeedEntityCollection } from "@/repositories/db-repository/feed-entity-repository";
 import { IPaperEntityCollection } from "@/repositories/db-repository/paper-entity-repository";
 
+import { CategorizerType } from "@/models/categorizer";
```
Comment (GeoffreyChen777) [inline app/renderer/ui/main-view/main-view.vue:12]: About the importing format.

Please remove the blank line above this and add a blank line below this.

**Comment by Rccubc** [comment]
OK

File: app/models/categorizer.ts (hunk 1 above)
Comment (GeoffreyChen777) [reply app/models/categorizer.ts:114]: OK, then rename it with `CategorizerMenuItem`

File: app/renderer/ui/main-view/data-view/paper-data-view.vue (hunk 3 above)
Comment (Rccubc) [reply app/renderer/ui/main-view/data-view/paper-data-view.vue:177]: I change this method to using a `Set`, code👇.  I'll benchmark it later.
```typescript
const getCategorizeList = (selectedList: number[]) => {
  let categorizeIdSet = new Set<string>();
  let categorizeList: CategorizeItem[] = [];
  selectedList.forEach((index) => {
    const paper = paperEntities.value[index];
    paper.folders.forEach((folder) => {
      if (!categorizeIdSet.has(`${folder._id}`)) {
        categorizeIdSet.add(`${folder._id}`);
        categorizeList.push({
          type: CategorizerType.PaperFolder,
          name: folder.name,
          id: folder._id,
        } as CategorizeItem);
      }
    });
    paper.tags.forEach((tag) => {
      if (!categorizeIdSet.has(`${tag._id}`)) {
        categorizeIdSet.add(`${tag._id}`);
        categorizeList.push({
          type: CategorizerType.PaperTag,
          name: tag.name,
          id: tag._id,
        } as CategorizeItem);
      }
    });
  });
  
  return categorizeList;
};
```

File: app/models/categorizer.ts (hunk 1 above)
Comment (Rccubc) [reply app/models/categorizer.ts:114]: rename and put it in `categorizer.ts`, right?

File: app/renderer/ui/main-view/data-view/paper-data-view.vue (hunk 3 above)
Comment (Rccubc) [reply app/renderer/ui/main-view/data-view/paper-data-view.vue:177]: bench using simple generated `paperEntities`, 5000 paperEntities, 40% folder/tag repeating, average of 10 times, the result is 0.8~1.2 ms

image


[sandbox](https://codesandbox.io/p/sandbox/bench-getcategorylist-pwk542?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clu1hiqj70006356gya2km9bz%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clu1hiqj70002356g5xwmp6cg%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clu1hiqj70003356gpetm2oar%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clu1hiqj70005356gcoehk9m8%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clu1hiqj70002356g5xwmp6cg%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clu1hiqj60001356g2ska2497%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.ts%2522%257D%255D%252C%2522id%2522%253A%2522clu1hiqj70002356g5xwmp6cg%2522%252C%2522activeTabId%2522%253A%2522clu1hiqj60001356g2ska2497%2522%257D%252C%2522clu1hiqj70005356gcoehk9m8%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clu1hiqj70004356ga4uzm0us%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clu1hiqj70005356gcoehk9m8%2522%252C%2522activeTabId%2522%253A%2522clu1hiqj70004356ga4uzm0us%2522%257D%252C%2522clu1hiqj70003356gpetm2oar%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clu1hiqj70003356gpetm2oar%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Comment by Rccubc** [comment]
Please check it.

**Review by GeoffreyChen777 — APPROVED** [review]

**Pull request merged**